File: apps/user-mobile/modules/chime-sdk/ios/MeetingObserver.swift
import AmazonChimeSDK

/// Observer for audio/video session events
class ChimeMeetingObserver: AudioVideoObserver, RealtimeObserver {
    private weak var module: ChimeSdkModule?

    init(module: ChimeSdkModule) {
        self.module = module
    }

    // MARK: - AudioVideoObserver

    func audioSessionDidStartConnecting(reconnecting: Bool) {
        if !reconnecting {
            module?.sendEvent([
                "type": "meetingStarted"
            ])
        }
    }

    func audioSessionDidStart(reconnecting: Bool) {
        // Session connected
    }

    func audioSessionDidDrop() {
        // Audio session dropped - will attempt to reconnect
    }

    func audioSessionDidStopWithStatus(sessionStatus: MeetingSessionStatus) {
        let statusCode = sessionStatus.statusCode

        switch statusCode {
        case .ok:
            module?.sendEvent([
                "type": "meetingEnded"
            ])
        default:
            module?.sendEvent([
                "type": "meetingFailed",
                "error": statusCode.description
            ])
        }
    }

    func audioSessionDidCancelReconnect() {
        module?.sendEvent([
            "type": "meetingFailed",
            "error": "Reconnection cancelled"
        ])
    }

    func connectionDidRecover() {
        // Connection recovered after temporary loss
    }

    func connectionDidBecomePoor() {
        // Connection quality degraded
    }

    func remoteVideoSourcesDidBecomeAvailable(sources: [RemoteVideoSource]) {
        // Remote video sources available
    }

    func remoteVideoSourcesDidBecomeUnavailable(sources: [RemoteVideoSource]) {
        // Remote video sources unavailable
    }

    func cameraSendAvailabilityDidChange(available: Bool) {
        // Camera send availability changed
    }

    func videoSessionDidStartConnecting() {
        // Video session connecting
    }

    func videoSessionDidStartWithStatus(sessionStatus: MeetingSessionStatus) {
        // Video session started
    }

    func videoSessionDidStopWithStatus(sessionStatus: MeetingSessionStatus) {
        // Video session stopped
    }

    // MARK: - RealtimeObserver

    func volumeDidChange(volumeUpdates: [VolumeUpdate]) {
        // Volume updates - can be used for audio level indicators
    }

    func signalStrengthDidChange(signalUpdates: [SignalUpdate]) {
        // Signal strength updates
    }

    func attendeesDidJoin(attendeeInfo: [AttendeeInfo]) {
        for info in attendeeInfo {
            module?.sendEvent([
                "type": "attendeeJoined",
                "attendeeId": info.attendeeId,
                "externalUserId": info.externalUserId
            ])
        }
    }

    func attendeesDidLeave(attendeeInfo: [AttendeeInfo]) {
        for info in attendeeInfo {
            module?.sendEvent([
                "type": "attendeeLeft",
                "attendeeId": info.attendeeId
            ])
        }
    }

    func attendeesDidDrop(attendeeInfo: [AttendeeInfo]) {
        for info in attendeeInfo {
            module?.sendEvent([
                "type": "attendeeLeft",
                "attendeeId": info.attendeeId
            ])
        }
    }

    func attendeesDidMute(attendeeInfo: [AttendeeInfo]) {
        for info in attendeeInfo {
            module?.sendEvent([
                "type": "attendeeMuteChanged",
                "attendeeId": info.attendeeId,
                "isMuted": true
            ])
        }
    }

    func attendeesDidUnmute(attendeeInfo: [AttendeeInfo]) {
        for info in attendeeInfo {
            module?.sendEvent([
                "type": "attendeeMuteChanged",
                "attendeeId": info.attendeeId,
                "isMuted": false
            ])
        }
    }
}

/// Observer for video tile events
class ChimeVideoTileObserver: VideoTileObserver {
    private weak var module: ChimeSdkModule?
    var videoTiles: [VideoTileState] = []

    init(module: ChimeSdkModule) {
        self.module = module
    }

    func videoTileDidAdd(tileState: VideoTileState) {
        videoTiles.append(tileState)

        module?.sendEvent([
            "type": "videoTileAdded",
            "tileState": [
                "tileId": tileState.tileId,
                "attendeeId": tileState.attendeeId ?? "",
                "isLocalTile": tileState.isLocalTile,
                "isContent": tileState.isContent,
                "pauseState": tileState.pauseState.rawValue,
                "videoStreamContentWidth": tileState.videoStreamContentWidth,
                "videoStreamContentHeight": tileState.videoStreamContentHeight
            ]
        ])
    }

    func videoTileDidRemove(tileState: VideoTileState) {
        videoTiles.removeAll { $0.tileId == tileState.tileId }

        module?.sendEvent([
            "type": "videoTileRemoved",
            "tileId": tileState.tileId
        ])
    }

    func videoTileDidPause(tileState: VideoTileState) {
        if let index = videoTiles.firstIndex(where: { $0.tileId == tileState.tileId }) {
            videoTiles[index] = tileState
        }

        module?.sendEvent([
            "type": "videoTilePaused",
            "tileId": tileState.tileId
        ])
    }

    func videoTileDidResume(tileState: VideoTileState) {
        if let index = videoTiles.firstIndex(where: { $0.tileId == tileState.tileId }) {
            videoTiles[index] = tileState
        }

        module?.sendEvent([
            "type": "videoTileResumed",
            "tileId": tileState.tileId
        ])
    }

    func videoTileSizeDidChange(tileState: VideoTileState) {
        if let index = videoTiles.firstIndex(where: { $0.tileId == tileState.tileId }) {
            videoTiles[index] = tileState
        }
    }
}

/// Observer for active speaker detection
class ChimeActiveSpeakerObserver: ActiveSpeakerObserver {
    private weak var module: ChimeSdkModule?

    var observerId: String {
        return "ChimeActiveSpeakerObserver"
    }

    init(module: ChimeSdkModule) {
        self.module = module
    }

    func activeSpeakerDidDetect(attendeeInfo: [AttendeeInfo]) {
        let attendeeIds = attendeeInfo.map { $0.attendeeId }
        module?.sendEvent([
            "type": "activeSpeakerChanged",
            "attendeeIds": attendeeIds
        ])
    }

    func activeSpeakerScoreDidChange(scores: [AttendeeInfo: Double]) {
        // Score updates for speaker detection
    }
}

File: apps/user-mobile/modules/chime-sdk/android/src/main/java/expo/modules/chimesdk/MeetingObserver.kt
package expo.modules.chimesdk

import com.amazonaws.services.chime.sdk.meetings.audiovideo.AudioVideoObserver
import com.amazonaws.services.chime.sdk.meetings.audiovideo.video.VideoTileObserver
import com.amazonaws.services.chime.sdk.meetings.audiovideo.video.VideoTileState
import com.amazonaws.services.chime.sdk.meetings.realtime.RealtimeObserver
import com.amazonaws.services.chime.sdk.meetings.session.MeetingSessionStatus
import com.amazonaws.services.chime.sdk.meetings.session.MeetingSessionStatusCode
import com.amazonaws.services.chime.sdk.meetings.audiovideo.AttendeeInfo
import com.amazonaws.services.chime.sdk.meetings.audiovideo.SignalUpdate
import com.amazonaws.services.chime.sdk.meetings.audiovideo.VolumeUpdate
import com.amazonaws.services.chime.sdk.meetings.audiovideo.video.VideoPauseState
import com.amazonaws.services.chime.sdk.meetings.audiovideo.audio.activespeakerdetector.ActiveSpeakerObserver
import com.amazonaws.services.chime.sdk.meetings.audiovideo.audio.activespeakerpolicy.DefaultActiveSpeakerPolicy

/**
 * Observer for audio/video session events
 */
class ChimeMeetingObserver(private val module: ChimeSdkModule) : AudioVideoObserver, RealtimeObserver {

    // AudioVideoObserver

    override fun onAudioSessionStartedConnecting(reconnecting: Boolean) {
        if (!reconnecting) {
            module.sendEvent(mapOf(
                "type" to "meetingStarted"
            ))
        }
    }

    override fun onAudioSessionStarted(reconnecting: Boolean) {
        // Session connected
    }

    override fun onAudioSessionDropped() {
        // Audio session dropped - will attempt to reconnect
    }

    override fun onAudioSessionStopped(sessionStatus: MeetingSessionStatus) {
        when (sessionStatus.statusCode) {
            MeetingSessionStatusCode.OK -> {
                module.sendEvent(mapOf(
                    "type" to "meetingEnded"
                ))
            }
            else -> {
                module.sendEvent(mapOf(
                    "type" to "meetingFailed",
                    "error" to sessionStatus.statusCode.name
                ))
            }
        }
    }

    override fun onAudioSessionCancelledReconnect() {
        module.sendEvent(mapOf(
            "type" to "meetingFailed",
            "error" to "Reconnection cancelled"
        ))
    }

    override fun onConnectionRecovered() {
        // Connection recovered after temporary loss
    }

    override fun onConnectionBecamePoor() {
        // Connection quality degraded
    }

    override fun onVideoSessionStartedConnecting() {
        // Video session connecting
    }

    override fun onVideoSessionStarted(sessionStatus: MeetingSessionStatus) {
        // Video session started
    }

    override fun onVideoSessionStopped(sessionStatus: MeetingSessionStatus) {
        // Video session stopped
    }

    // RealtimeObserver

    override fun onVolumeChanged(volumeUpdates: Array<VolumeUpdate>) {
        // Volume updates - can be used for audio level indicators
    }

    override fun onSignalStrengthChanged(signalUpdates: Array<SignalUpdate>) {
        // Signal strength updates
    }

    override fun onAttendeesJoined(attendeeInfo: Array<AttendeeInfo>) {
        for (info in attendeeInfo) {
            module.sendEvent(mapOf(
                "type" to "attendeeJoined",
                "attendeeId" to info.attendeeId,
                "externalUserId" to info.externalUserId
            ))
        }
    }

    override fun onAttendeesLeft(attendeeInfo: Array<AttendeeInfo>) {
        for (info in attendeeInfo) {
            module.sendEvent(mapOf(
                "type" to "attendeeLeft",
                "attendeeId" to info.attendeeId
            ))
        }
    }

    override fun onAttendeesDropped(attendeeInfo: Array<AttendeeInfo>) {
        for (info in attendeeInfo) {
            module.sendEvent(mapOf(
                "type" to "attendeeLeft",
                "attendeeId" to info.attendeeId
            ))
        }
    }

    override fun onAttendeesMuted(attendeeInfo: Array<AttendeeInfo>) {
        for (info in attendeeInfo) {
            module.sendEvent(mapOf(
                "type" to "attendeeMuteChanged",
                "attendeeId" to info.attendeeId,
                "isMuted" to true
            ))
        }
    }

    override fun onAttendeesUnmuted(attendeeInfo: Array<AttendeeInfo>) {
        for (info in attendeeInfo) {
            module.sendEvent(mapOf(
                "type" to "attendeeMuteChanged",
                "attendeeId" to info.attendeeId,
                "isMuted" to false
            ))
        }
    }
}

/**
 * Observer for video tile events
 */
class ChimeVideoTileObserver(private val module: ChimeSdkModule) : VideoTileObserver {
    val videoTiles = mutableListOf<VideoTileState>()

    override fun onVideoTileAdded(tileState: VideoTileState) {
        videoTiles.add(tileState)

        module.sendEvent(mapOf(
            "type" to "videoTileAdded",
            "tileState" to mapOf(
                "tileId" to tileState.tileId,
                "attendeeId" to (tileState.attendeeId ?: ""),
                "isLocalTile" to tileState.isLocalTile,
                "isContent" to tileState.isContent,
                "pauseState" to tileState.pauseState.ordinal,
                "videoStreamContentWidth" to tileState.videoStreamContentWidth,
                "videoStreamContentHeight" to tileState.videoStreamContentHeight
            )
        ))
    }

    override fun onVideoTileRemoved(tileState: VideoTileState) {
        videoTiles.removeAll { it.tileId == tileState.tileId }

        module.sendEvent(mapOf(
            "type" to "videoTileRemoved",
            "tileId" to tileState.tileId
        ))
    }

    override fun onVideoTilePaused(tileState: VideoTileState) {
        val index = videoTiles.indexOfFirst { it.tileId == tileState.tileId }
        if (index >= 0) {
            videoTiles[index] = tileState
        }

        module.sendEvent(mapOf(
            "type" to "videoTilePaused",
            "tileId" to tileState.tileId
        ))
    }

    override fun onVideoTileResumed(tileState: VideoTileState) {
        val index = videoTiles.indexOfFirst { it.tileId == tileState.tileId }
        if (index >= 0) {
            videoTiles[index] = tileState
        }

        module.sendEvent(mapOf(
            "type" to "videoTileResumed",
            "tileId" to tileState.tileId
        ))
    }

    override fun onVideoTileSizeChanged(tileState: VideoTileState) {
        val index = videoTiles.indexOfFirst { it.tileId == tileState.tileId }
        if (index >= 0) {
            videoTiles[index] = tileState
        }
    }
}

/**
 * Observer for active speaker detection
 */
class ChimeActiveSpeakerObserver(private val module: ChimeSdkModule) : ActiveSpeakerObserver {

    override val observerId: String = "ChimeActiveSpeakerObserver"

    override fun onActiveSpeakerDetected(attendeeInfo: Array<AttendeeInfo>) {
        val attendeeIds = attendeeInfo.map { it.attendeeId }
        module.sendEvent(mapOf(
            "type" to "activeSpeakerChanged",
            "attendeeIds" to attendeeIds
        ))
    }

    override fun onActiveSpeakerScoreChanged(scores: Map<AttendeeInfo, Double>) {
        // Score updates for speaker detection
    }
}
